File: backend/app/provider/pixiv.py
import json
import os

import requests
import bs4

from app.provider.base import BookProvider
from app.model import (
    Author,
    TocEpisodeToken,
    BookMetadata,
    Episode,
)


def _create_headers():
    headers = {
        "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36",
    }
    phpsessid = os.environ.get("PIXIV_COOKIE_PHPSESSID")
    if phpsessid:
        headers["cookie"] = f"PHPSESSID={phpsessid}"
    return headers


class Pixiv(BookProvider):
    provider_id = "pixiv"
    lang = "jp"

    base_url = "https://www.pixiv.net"
    session = requests.Session()
    headers = _create_headers()

    def _fetch(self, url):
        res = self.session.get(
            self.base_url + url,
            headers=self.headers,
        )
        res.raise_for_status()
        return res

    @staticmethod
    def _build_authors(user_name: str, user_id: str):
        link = f"{Pixiv.base_url}/users/{user_id}"
        return [Author(name=user_name, link=link)]

    def _get_book_metadata(self, book_id: str) -> BookMetadata:
        if book_id.startswith("s"):
            return self._parse_metadata_single(book_id[1:])
        else:
            return self._parse_metadata_serial(book_id)

    def _parse_metadata_serial(self, book_id: str) -> BookMetadata:
        res = self._fetch(f"/ajax/novel/series/{book_id}/content_titles")
        toc = [
            TocEpisodeToken(title=obj["title"], episode_id=obj["id"])
            for obj in res.json()["body"]
        ]

        res = self._fetch(f"/ajax/novel/series/{book_id}")
        obj = res.json()["body"]

        return BookMetadata(
            title=obj["title"],
            authors=self._build_authors(obj["userName"], obj["userId"]),
            introduction=obj["caption"],
            toc=toc,
        )

    def _parse_metadata_single(self, episode_id: str) -> BookMetadata:
        res = self._fetch(f"/novel/show.php?id={episode_id}")
        soup = bs4.BeautifulSoup(res.text, "html.parser")

        obj = json.loads(soup.find("meta", {"id": "meta-preload-data"})["content"])
        obj = obj["novel"][episode_id]
        assert obj["seriesNavData"] is None

        return BookMetadata(
            title=obj["title"],
            authors=self._build_authors(obj["userName"], obj["userId"]),
            introduction=obj["description"].replace("<br />", "\n"),
            toc=[TocEpisodeToken(title="", episode_id=episode_id)],
        )

    def _get_episode(
        self,
        book_id: str,
        episode_id: str,
    ) -> Episode:
        res = self._fetch(f"/novel/show.php?id={episode_id}")
        soup = bs4.BeautifulSoup(res.text, "html.parser")

        obj = json.loads(soup.find("meta", {"id": "meta-preload-data"})["content"])
        obj = obj["novel"][episode_id]

        return Episode(paragraphs=obj["content"].splitlines())

File: src/main/kotlin/data/provider/providers/Pixiv.kt
package data.provider.providers

import data.provider.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive


class Pixiv : BookProvider {
    companion object {
        const val id = "pixiv"
    }

    private suspend fun get(url: String): HttpResponse {
        return client.get(url) {
            headers {
                append(
                    HttpHeaders.UserAgent,
                    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36"
                )
            }
            System.getenv("PIXIV_COOKIE_PHPSESSID")?.let {
                cookie(name = "PHPSESSID", value = it, domain = ".pixiv.net")
            }
        }
    }

    override suspend fun getRank(options: Map<String, String>): List<SBookListItem> {
        TODO("Not yet implemented")
    }

    override fun getMetadataUrl(bookId: String): String {
        return if (bookId.startsWith("s")) {
            getEpisodeUrl(bookId, bookId.removePrefix("s"))
        } else {
            "https://www.pixiv.net/novel/series/${bookId}"
        }
    }

    override fun getEpisodeUrl(bookId: String, episodeId: String): String {
        return "https://www.pixiv.net/novel/show.php?id=$episodeId"
    }

    override suspend fun getMetadata(bookId: String): SBookMetadata {
        if (bookId.startsWith("s")) {
            val episodeId = bookId.removePrefix("s")
            val doc = get(getEpisodeUrl(bookId, episodeId)).document()
            val jsonRaw = doc.selectFirst("meta#meta-preload-data")!!.attr("content")
            val obj = Json.parseToJsonElement(jsonRaw).jsonObject["novel"]!!.jsonObject[episodeId]!!.jsonObject
            assert(obj["seriesNavData"] != null)

            val title = obj["title"]!!.jsonPrimitive.content
            val author = SBookAuthor(
                name = obj["userName"]!!.jsonPrimitive.content,
                link = "https://www.pixiv.net/users/" + obj["userId"]!!.jsonPrimitive.content
            )
            val introduction = obj["description"]!!.jsonPrimitive.content.replace("<br />", "\n")

            return SBookMetadata(
                title = title,
                authors = listOf(author),
                introduction = introduction,
                toc = listOf(SBookTocItem(title = "无名", episodeId = episodeId)),
            )
        } else {
            val obj1 = get("https://www.pixiv.net/ajax/novel/series/$bookId").json()["body"]!!.jsonObject
            val title = obj1["title"]!!.jsonPrimitive.content
            val author = SBookAuthor(
                name = obj1["userName"]!!.jsonPrimitive.content,
                link = "https://www.pixiv.net/users/" + obj1["userId"]!!.jsonPrimitive.content
            )
            val introduction = obj1["caption"]!!.jsonPrimitive.content

            val obj2 = get("https://www.pixiv.net/ajax/novel/series/$bookId/content_titles").json()
            val toc = obj2["body"]!!.jsonArray
                .map { it.jsonObject }
                .map {
                    SBookTocItem(
                        title = it["title"]!!.jsonPrimitive.content,
                        episodeId = it["id"]!!.jsonPrimitive.content,
                    )
                }

            return SBookMetadata(
                title = title,
                authors = listOf(author),
                introduction = introduction,
                toc = toc,
            )
        }
    }

    override suspend fun getEpisode(bookId: String, episodeId: String): SBookEpisode {
        val doc = get(getEpisodeUrl(bookId, episodeId)).document()
        val jsonRaw = doc.selectFirst("meta#meta-preload-data")!!.attr("content")
        val obj = Json.parseToJsonElement(jsonRaw).jsonObject["novel"]!!.jsonObject[episodeId]!!.jsonObject
        return SBookEpisode(paragraphs = obj["content"]!!.jsonPrimitive.content.lines())
    }
}
